File: src/algorithm/leetcode/Problem101.java
package algorithm.leetcode;

public class Problem101 {

    public static boolean isSymmetric(TreeNode a, TreeNode b) {
        return (a == b) || (a != null && b != null && a.val == b.val &&
                isSymmetric(a.left, b.right) && isSymmetric(b.left, a.right));
    }

    public static boolean isSymmetric(TreeNode root) {
        return root == null || isSymmetric(root.left, root.right);
    }

    public static void main(String[] args) {
        TreeNode a = new TreeNode(1);
        TreeNode b = new TreeNode(2);
        a.left = b;
        isSymmetric(a);
    }
}

File: src/algorithm_cpp/problem101.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool mirror(TreeNode* left, TreeNode* right) {
        if (left == nullptr || right == nullptr) {
            return left == nullptr && right == nullptr;
        }
        if (left->val != right->val) {
            return false;
        }
        return mirror(left->right, right->left) && mirror(left->left, right->right);
    }

    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) return true;
        return mirror(root->left, root->right);
    }
};
